### ISSUE TITLE
Spike: automate move-candidate-drive for candidates on the full-to-capacity shared drive / move to store candidate files on S3

### ISSUE THREAD

**Issue by samschlicht** [talentbeyondboundaries/tbbtalentv2#312]
Triggered by the issue experienced when one of our Google Shared Drives ran out of object limit (400,000). We have a reasonable but manual process for checking capacity and making a new drive when needed, but S3 might be the solution to remove that requirement.

We'll need to scope the amount of work and decide whether it's something we want to do and when. A low-cost alternative would be to automate a warning when the Drive is at, say, 80% capacity. For now, there has to be a manual process to add the new drive, since the AWS task definition has to change to reflect the new Drive and folder ID — but that could change once we're reliably terraforming.

**Comment by samschlicht** [comment]
Ultimate goal is to automate this, but being mindful of time before release, I wanted to see if I could buy us some breathing space!

I found a quick way of moving 1000s of candidates at once.

Using Google Drive extension for Mac (quicker than web-browser Google Drive, but that'll work too):
1. Shared Drives > CandidateData > Candidate Folders
2. From the later candidate folders (because their names only contain candidate #s), block-select a big batch and copy
4. Open a Google Sheet and paste them in, download as CSV
5. In the TC create a new list, import the CSV — you'll have a list of all the candidates whose folders you copied in step 2
6. Copy the ID from the list URL
7. Settings > Admin API > execute this call, appending the list ID, e.g. `move-candidates-drive/7334 `
8. As the method doc in SystemAdminApi.java notes, you'll get a timeout error — but the moves will happen

**Comment by camerojo** [comment]
I think that the ultimate solution to this problem is to move most candidate data off the Google Drive and on to Amazon S3.

Amazon S3 has limits on its "buckets" (100-1000) but no limitation on the number of files stored in a bucket. And there is no performance penalty putting ALL files in a single bucket (if needed).
https://docs.aws.amazon.com/AmazonS3/latest/userguide/BucketRestrictions.html

Unlike Google Drive, S3 does not support any file structure - ie no concept of folders as Google has. We would have to impose that structure from the TC - so for example a candidate "folder" would correspond to a collection of links to S3 files on the TC database, associated with the relevant candidate. This wouldn't be a bad thing anyway - allowing us to easily display the contents of that "folder" on the TC User interface - for example as a tab on the TC candidate display.
In return we avoid the file limits on Google Drives.

I expect that we could keep the Google Drive for generating, for example, sheets summarizing Candidate lists that we can share with external parties. But for the storage of uploaded candidate documents - like passports, Cvs, etc - we should move over to using S3.

**Comment by samschlicht** [comment]
Confirming that S3 storage would run to around US$8 per month at our current level, so cost is not really an issue.

The main issue is the amount of work and whether this is a priority at the moment. I think the proposition is fairly straightforward: S3 is a market leader, reasonably priced, and as John notes above there's no restriction on the capacity of an individual bucket, though I would emphasise his point about S3's lack of bucket structure, which is not a big problem but will require a different approach; Google Drive is free and working okay apart from the object limit, which is non-negotiable (I've checked).

Looking at the code, there's a lot that would need to change: I would recommend for now that we perhaps add a monitoring and alert automation, which precludes the possibility that we take our eye off the ball and hit the limit.  We could even go a bit further and create the new drives. But, as I mentioned, until we move to dependably terraforming prod, there would be the manual step of updating the AWS task definition variables and redeploying.

My reasoning is thus: at our current rate, we have to make a new drive roughly every 8 months. The process takes 30 mins to an hour. It's hard to estimate, but switching to S3 might use at least a week's development time (between design, implementation and review), and create a fair amount of exposure to bugs. That's all fine and manageable, but to my mind not a priority right now, and the pay-off would take a long time to realise. I'll put it to the group though.

NB: I haven't got into the technical side of implementing S3 too much — [the API doc](https://docs.aws.amazon.com/AmazonS3/latest/API/API_Operations_Amazon_Simple_Storage_Service.html) seems pretty straightforward and I think we can count on it being a logical and reasonable process; critically, the limits on buckets pose no issue.

That being said, I'm happy to scope this in more detail if we feel that it's a priority!

**Comment by samschlicht** [comment]
Not a game-changer, but just noting that the item cap on Google Shared Drives has gone up to 500,000.

<img width="336" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMkZ6Q0E9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--fa1488cc981e55efa376b8ebd7c2f8dae31fbcfe/Screenshot%202025-01-16%20at%204.36.58%E2%80%AFpm.png" alt="Screenshot 2025-01-16 at 4.36.58 pm.png" />

**Comment by samschlicht** [comment]
Agreed with John that this should be a candidate for the next release. It's a big lift and not a priority right now.
